### PR TITLE
Add an "agreement" challenge

### PR DESCRIPTION
**Pull request by bifurcation**
The [CABF Baseline Requirements (Section 10.3.1)](https://cabforum.org/wp-content/uploads/BRv1.2.3.pdf) require CAs to obtain agreement with a "Subscriber Agreement" or "Terms of Use" before they will issue a certificate.  We can reflect this requirement in ACME with a new challenge type that allows an applicant for authorization to agree to terms for the issuance of certs under that authorization.

### CODE REVIEW

**Comment by bifurcation** [comment]
This PR addresses Issue #45

**Comment by jsha** [comment]
I think this makes more sense as a parameter to POST /new-reg. That is, when you register your account key with the CA, you agree to terms of service then, and they cover all certificates issued under that account key.

**Comment by martinthomson** [comment]
I agree with @jsha here.  Otherwise, building a challenge each time would be hard.  You might like to note that an update to the terms of use should also cause any existing acceptance for that account to be invalidated.

**Comment by bifurcation** [comment]
I would really like to keep the registration to the client POSTing information about itself.  How about this:
* Response to registration includes "Link: <terms.html>;rel=terms-of-service"
* If the client agrees, he sends an updated POST with an attribute "agreed: <hash-of-ToS>"

**Comment by martinthomson** [comment]
@bifurcation that looks fine.  I was thinking that the registration could link to the terms of service and the client could indicate the specific version that they agreed to somehow.  But a boolean is probably OK as a lower bar.

If you want a specific version, you could have a different ToS link for each and require the client to indicate *which* ToS they agree to.  (You could also use etags to identify specific versions of a single URI, but that's kludgy).  The unique link works more easily though I think.

**Comment by bifurcation** [comment]
How about a middle path: The client just says "Here's the URI of the terms I agreed to".  That way the client can indicate which version it agreed to, and the server can control the versioning by assigning different URIs to different versions, but there's not a whole lot of details baked into the protocol.  

I've updated the PR with this.  If nobody objects, I'll go ahead and merge.

File: draft-barnes-acme.md [hunk 1]
```diff
@@ -344,12 +347,16 @@ provided by the client.
 The server creates a registration object with the included contact information.  The "key" element of the registration is set to the public key used to verify the JWS (i.e., the "jwk" element of the JWS header).  The server also provides a random
 recovery token.  The server returns this registration object in a 201 (Created) response, with the registration URI in a Location header field.  The server may also indicate its new-authorization URI using the "next" link relation.
 
+If the server wishes to present the client with terms under which the ACME service is to be used, it may indicate the URI where such terms can be accessed in a Link header with link relation "terms-of-service".  As noted above, the client may indicate its
```
Comment (martinthomson) [inline draft-barnes-acme.md]: "may", or MUST?

Comment (martinthomson) [inline draft-barnes-acme.md]: I guess the point is that the client might reference an old, but perfectly acceptable version.  The server has to decide whether it needs to reject the request.

Is there any automated way to signal that the agreement was not current or correct?

**Comment by martinthomson** [comment]
BTW, I think that you should merge without addressing those concerns, though I'd request that you open an issue with them if you don't (or maybe you can reject the comments).

File: draft-barnes-acme.md [hunk 2]
```diff
@@ -344,12 +347,16 @@ provided by the client.
...
+agreement with these terms by updating its registration to include the "agreement" field, with the terms URI as its value.
+
 ~~~~~~~~~~
 
 HTTP/1.1 201 Created
 Content-Type: application/json
 Location: https://example.com/reg/asdf
 Link: <https://example.com/acme/new-authz>;rel="next"
+Link: <https://example.com/acme/terms>;rel="terms-of-service"
```
Comment (martinthomson) [inline draft-barnes-acme.md]: You will need to define a new link relation type, I think.

Comment (bifurcation) [inline draft-barnes-acme.md]: Why is "terms-of-service" not sufficient?  It's already defined, and seems to fit well enough.  I'm going to go ahead and merge, and we can handle this separately if we need to.

Comment (martinthomson) [inline draft-barnes-acme.md]: Yeah, I checked 5988, not the registry.  You are right.

**Pull request merged**